File: crabzilla-core/src/main/java/io/github/crabzilla/core/entity/StateTransitionsTracker.java
package io.github.crabzilla.core.entity;

import io.github.crabzilla.core.DomainEvent;
import lombok.NonNull;
import lombok.val;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;

public class StateTransitionsTracker<A extends Entity>  {

  private final Snapshot<A> originalSnapshot;
  private final BiFunction<DomainEvent, A, A> applyEventsFn;
  private final List<StateTransition<A>> stateTransitions = new ArrayList<>();

  public StateTransitionsTracker(@NonNull Snapshot<A> originalSnapshot,
                                 @NonNull BiFunction<DomainEvent, A, A> applyEventsFn) {
    this.originalSnapshot = originalSnapshot;
    this.applyEventsFn = applyEventsFn;
  }

  private StateTransitionsTracker<A> applyEvents(@NonNull List<DomainEvent> events) {
    events.forEach(e -> {
      val newInstance = applyEventsFn.apply(e, currentState());
      stateTransitions.add(new StateTransition<>(newInstance, e));
    });
    return this;
  }

  public StateTransitionsTracker<A> applyEvents(@NonNull Function<A, List<DomainEvent>> function) {
    val newEvents = function.apply(currentState());
    return applyEvents(newEvents);
  }

  public List<DomainEvent> collectEvents() {
    return stateTransitions.stream().map(t -> t.afterThisEvent).collect(Collectors.toList());
  }

  public A currentState() {
    val current = isEmpty() ? originalSnapshot.getInstance() :
                              stateTransitions.get(stateTransitions.size() - 1).newInstance;
    return current;
  }

  public boolean isEmpty() {
    return stateTransitions.isEmpty();
  }

  public EntityUnitOfWork unitOfWorkFor(EntityCommand command) {
    return EntityUnitOfWork.unitOfWork(command, originalSnapshot.nextVersion(), collectEvents());
  }

  class StateTransition<T extends Entity> {
    private final T newInstance;
    private final DomainEvent afterThisEvent;

    StateTransition(@NonNull T newInstance, @NonNull DomainEvent afterThisEvent) {
      this.newInstance = newInstance;
      this.afterThisEvent = afterThisEvent;
    }
  }
}

File: crabzilla-core/src/main/java/io/github/crabzilla/StateTransitionsTracker.kt
package io.github.crabzilla

import java.util.*

class StateTransitionsTracker<A : Entity>(private val originalSnapshot: Snapshot<A>,
                                          private val applyEventsFn: (DomainEvent, A) -> A) {
  private val stateTransitions = ArrayList<StateTransition<A>>()

  inline fun applyEvents(aggregateRootMethodFn: (A) -> List<DomainEvent>): StateTransitionsTracker<A> {
    val newEvents = aggregateRootMethodFn.invoke(currentState())
    return applyEvents(newEvents)
  }

  fun collectEvents(): List<DomainEvent> {
    return stateTransitions.map { t -> t.afterThisEvent }.toList()
  }

  fun currentState(): A {
    return if (isEmpty) originalSnapshot.instance else stateTransitions[stateTransitions.size - 1].newInstance
  }

  val isEmpty: Boolean
    get() = stateTransitions.isEmpty()

  fun applyEvents(events: List<DomainEvent>): StateTransitionsTracker<A> {
    events.forEach { e ->
      val newInstance = applyEventsFn.invoke(e, currentState())
      stateTransitions.add(StateTransition(newInstance, e))
    }
    return this
  }
  internal inner class StateTransition<out T : Entity>(val newInstance: T, val afterThisEvent: DomainEvent)
}
